packages/libs/eda/src/lib/core/components/visualizations/implementations/VolcanoPlotVisualization.tsx

```diff
@@ -148,6 +159,146 @@ function VolcanoPlotViz(props: VisualizationProps<Options>) {
+          max: dataXMax + (dataXMax - dataXMin) * AXIS_PADDING_FACTOR,
+        };
+      } else {
+        return { min: 0, max: 0 };
+      }
+    }
+  }, [data.value, xAxisRange, rawDataMinMaxValues]);
+
+  // NOTE: this state may be unnecessary depending on how we implement user-controlled axis ranges
+  const [yAxisRange, setYAxisRange] =
+    useState<NumberRange | undefined>(undefined);
+  const dependentAxisRange = useMemo(() => {
+    if (!data.value) return undefined;
+    if (yAxisRange) {
+      return yAxisRange;
+    } else {
+      const {
+        y: { min: dataYMin, max: dataYMax },
+      } = rawDataMinMaxValues;
+      if (dataYMin && dataYMax) {
+        // Standard volcano plots have -log10(raw p value) as the y axis
+        const yAxisMin = -Math.log10(dataYMax);
+        const yAxisMax = -Math.log10(dataYMin);
+        // Add a little padding to prevent clipping the glyph representing the extreme points
+        return {
+          min: yAxisMin - (yAxisMax - yAxisMin) * AXIS_PADDING_FACTOR,
+          max: yAxisMax + (yAxisMax - yAxisMin) * AXIS_PADDING_FACTOR,
+        };
+      } else {
+        return { min: 0, max: 0 };
+      }
+    }
+  }, [data.value, yAxisRange, rawDataMinMaxValues]);
+
+  const significanceThreshold =
+    vizConfig.significanceThreshold ?? DEFAULT_SIG_THRESHOLD;
+  const log2FoldChangeThreshold =
+    vizConfig.log2FoldChangeThreshold ?? DEFAULT_FC_THRESHOLD;
+
+  /**
+   * Let's filter out data that falls outside of the plot axis ranges and then
+   * assign a significance color to the visible data
+   * This version of the data will get passed to the VolcanoPlot component
+   */
+  const finalData = useMemo(() => {
+    if (data.value && independentAxisRange && dependentAxisRange) {
+      // Only return data if the points fall within the specified range! Otherwise they'll show up on the plot.
+      return data.value
+        .filter((d) => {
+          const log2foldChange = Number(d?.log2foldChange);
+          const transformedPValue = -Math.log10(Number(d?.pValue));
+          return (
+            log2foldChange <= independentAxisRange.max &&
+            log2foldChange >= independentAxisRange.min &&
+            transformedPValue <= dependentAxisRange.max &&
+            transformedPValue >= dependentAxisRange.min
+          );
+        })
+        .map((d) => ({
+          ...d,
+          significanceColor: assignSignificanceColor(
+            Number(d.log2foldChange),
+            Number(d.pValue),
+            significanceThreshold,
+            log2FoldChangeThreshold,
+            significanceColors
+          ),
+        }));
+    }
+  }, [
+    data.value,
+    independentAxisRange,
+    dependentAxisRange,
+    significanceThreshold,
+    log2FoldChangeThreshold,
+    significanceColors,
+  ]);
+
+  // TODO: Given that I cannot figure out how to type this correctly, I suspect there's a simpler, cleaner
+  // way to get the counts.
```

Comment:
the other way i could think of is `significanceColors.map(color, finalData.filter(blah = color).length())` but i think your way is more performant?

Reply:
@dmfalke Do you have thoughts about this? I've found myself reaching for `reduce` a lot lately and I struggle to type it every time, which leads me to believe it's a code smell.

Reply:
I'm looking at this now...

Reply:
Thanks for the feedback. Yours is far more digestible, so I'll go with that approach. I've been having too much fun with `reduce` lately! 😆